plugin-server/src/cdp/hog-masker.ts

```diff
@@ -0,0 +1,131 @@
+import { exec } from '@posthog/hogvm'
+import { createHash } from 'crypto'
+
+import { CdpRedis } from './redis'
+import { HogFunctionInvocationGlobals, HogFunctionType } from './types'
+
+export const BASE_REDIS_KEY = process.env.NODE_ENV == 'test' ? '@posthog-test/hog-masker' : '@posthog/hog-masker'
+const REDIS_KEY_TOKENS = `${BASE_REDIS_KEY}/mask`
+
+// NOTE: These are controlled via the api so are more of a sanity fallback
+const MASKER_MAX_TTL = 60 * 60 * 24
+const MASKER_MIN_TTL = 1
+
+type MaskContext = {
+    hogFunctionId: string
+    hash: string
+    increment: number
+    ttl: number
+    allowedExecutions: number
+    threshold: number | null
+}
+
+type HogInvocationContext = {
+    globals: HogFunctionInvocationGlobals
+    hogFunction: HogFunctionType
+}
+
+type HogInvocationContextWithMasker = HogInvocationContext & {
+    masker?: MaskContext
+}
+
+/**
+ * HogMasker
+ *
+ * Responsible for determining if a function is "masked" or not based on the function configuration
+ */
+
+// Hog masker is meant to be done per batch
+export class HogMasker {
+    constructor(private redis: CdpRedis) {}
+
+    public async filterByMasking(invocations: HogInvocationContext[]): Promise<{
+        masked: HogInvocationContext[]
+        notMasked: HogInvocationContext[]
+    }> {
+        const invocationsWithMasker: HogInvocationContextWithMasker[] = [...invocations]
+        const masks: Record<string, MaskContext> = {}
+
+        // We find all functions that have a mask and we load their masking from redis
+        invocationsWithMasker.forEach((item) => {
+            if (item.hogFunction.masking) {
+                // TODO: Catch errors
+                const value = exec(item.hogFunction.masking.bytecode, {
+                    globals: item.globals,
+                    timeout: 50,
+                    maxAsyncSteps: 0,
+                })
+                // What to do if it is null....
+                const hash = createHash('md5').update(String(value.result)).digest('hex').substring(0, 32)
+                const hashKey = `${item.hogFunction.id}:${hash}`
+                masks[hashKey] = masks[hashKey] || {
+                    hash,
+                    hogFunctionId: item.hogFunction.id,
+                    increment: 0,
+                    ttl: Math.max(
+                        MASKER_MIN_TTL,
+                        Math.min(MASKER_MAX_TTL, item.hogFunction.masking.ttl ?? MASKER_MAX_TTL)
+                    ),
+                    threshold: item.hogFunction.masking.threshold,
+                    allowedExecutions: 0,
+                }
+
+                masks[hashKey]!.increment++
+                item.masker = masks[hashKey]
+            }
+        })
+
+        if (Object.keys(masks).length === 0) {
+            return { masked: [], notMasked: invocations }
+        }
+
+        // Load from redis returning the value - this allows us to compare - if the value is the same then we allow an invocation
```

Comment:
As an outsider, I'm a little fuzzy at exactly what's going on below (I mean I get the purpose and result, just not precisely the logic), and this comment didn't really help.

Reply:
Fair point. The comment is a bit outdated anyways as the solution evolved.